File: infrastructure/persistence/app_state_repository.py
from __future__ import annotations

import json
from datetime import datetime

from domain.models.app_state import AppState
from domain.models.preferences import UserPreferences
from infrastructure.persistence.file_write_utils import write_json_atomic
from infrastructure.persistence.runtime_paths_provider import RuntimePathsProvider


class AppStateRepository:
    _KEY_MINIMIZE_TO_TRAY_ON_CLOSE = "minimize_to_tray_on_close"
    _KEY_AUTO_SWITCH_ON_TRAFFIC_EXHAUSTED = "auto_switch_account_on_traffic_exhausted"
    _KEY_RECENT_ACCOUNT_IDS = "recent_account_ids"
    _MAX_RECENT_ACCOUNT_IDS = 20

    def __init__(self, paths: RuntimePathsProvider) -> None:
        self._paths = paths
        self._state_path = paths.get_app_state_path()

    def load_state(self) -> AppState:
        if self._state_path.exists():
            return self._load_from_path(self._state_path)

        legacy_path = self._paths.get_legacy_app_state_path()
        if legacy_path.exists():
            return self._load_from_path(legacy_path)
        return AppState()

    def load_preferences(self) -> UserPreferences:
        payload = self._load_payload()
        return UserPreferences(
            minimize_to_tray_on_close=bool(payload.get(self._KEY_MINIMIZE_TO_TRAY_ON_CLOSE, False)),
            auto_switch_account_on_traffic_exhausted=bool(payload.get(self._KEY_AUTO_SWITCH_ON_TRAFFIC_EXHAUSTED, False)),
            recent_account_ids=self._normalize_recent_account_ids(payload.get(self._KEY_RECENT_ACCOUNT_IDS, [])),
        )

    def save_state(self, state: AppState) -> None:
        payload = self._load_payload()
        payload.update({
            "last_login_time": state.last_login_time.isoformat() if state.last_login_time else "",
            "last_quota_refresh_time": (
                state.last_quota_refresh_time.isoformat() if state.last_quota_refresh_time else ""
            ),
            "last_login_result": state.last_login_result,
            "last_login_message": state.last_login_message,
        })
        write_json_atomic(self._state_path, payload)

    def set_minimize_to_tray_on_close(self, enabled: bool) -> None:
        payload = self._load_payload()
        payload[self._KEY_MINIMIZE_TO_TRAY_ON_CLOSE] = bool(enabled)
        write_json_atomic(self._state_path, payload)

    def set_auto_switch_on_traffic_exhausted(self, enabled: bool) -> None:
        payload = self._load_payload()
        payload[self._KEY_AUTO_SWITCH_ON_TRAFFIC_EXHAUSTED] = bool(enabled)
        write_json_atomic(self._state_path, payload)

    def mark_account_used(self, account_id: str) -> list[str]:
        account_id_text = account_id.strip()
        if not account_id_text:
            return self.get_recent_account_ids()

        reordered = [account_id_text]
        for existed_id in self.get_recent_account_ids():
            if existed_id != account_id_text:
                reordered.append(existed_id)

        self._save_recent_account_ids(reordered)
        return reordered

    def prune_recent_account_ids(self, valid_account_ids: set[str]) -> list[str]:
        kept = [account_id for account_id in self.get_recent_account_ids() if account_id in valid_account_ids]
        self._save_recent_account_ids(kept)
        return kept

    def get_recent_account_ids(self) -> list[str]:
        payload = self._load_payload()
        return self._normalize_recent_account_ids(payload.get(self._KEY_RECENT_ACCOUNT_IDS, []))

    @staticmethod
    def _load_from_path(path) -> AppState:
        try:
            payload = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return AppState()

        raw_last_login_time = str(payload.get("last_login_time", "")).strip()
        last_login_time: datetime | None = None
        if raw_last_login_time:
            try:
                last_login_time = datetime.fromisoformat(raw_last_login_time)
            except ValueError:
                last_login_time = None

        raw_last_quota_refresh_time = str(payload.get("last_quota_refresh_time", "")).strip()
        last_quota_refresh_time: datetime | None = None
        if raw_last_quota_refresh_time:
            try:
                last_quota_refresh_time = datetime.fromisoformat(raw_last_quota_refresh_time)
            except ValueError:
                last_quota_refresh_time = None

        return AppState(
            last_login_time=last_login_time,
            last_quota_refresh_time=last_quota_refresh_time,
            last_login_result=str(payload.get("last_login_result", "未执行")).strip() or "未执行",
            last_login_message=str(payload.get("last_login_message", "-")).strip() or "-",
        )

    def _load_payload(self) -> dict:
        if self._state_path.exists():
            return self._safe_read_payload(self._state_path)
        legacy_path = self._paths.get_legacy_app_state_path()
        if legacy_path.exists():
            return self._safe_read_payload(legacy_path)
        return {}

    @staticmethod
    def _safe_read_payload(path) -> dict:
        try:
            payload = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return {}
        return payload if isinstance(payload, dict) else {}

    def _save_recent_account_ids(self, account_ids: list[str]) -> None:
        payload = self._load_payload()
        payload[self._KEY_RECENT_ACCOUNT_IDS] = self._normalize_recent_account_ids(account_ids)
        write_json_atomic(self._state_path, payload)

    def _normalize_recent_account_ids(self, raw_value: object) -> list[str]:
        if isinstance(raw_value, str):
            candidates = [raw_value]
        elif isinstance(raw_value, list):
            candidates = raw_value
        else:
            candidates = []

        normalized: list[str] = []
        seen: set[str] = set()
        for item in candidates:
            account_id = str(item).strip()
            if not account_id or account_id in seen:
                continue
            seen.add(account_id)
            normalized.append(account_id)
            if len(normalized) >= self._MAX_RECENT_ACCOUNT_IDS:
                break
        return normalized

File: src-tauri/src/infrastructure/persistence/app_state_repository.rs
use chrono::{DateTime, Local};
use serde::{Deserialize, Serialize};

use crate::application::error::AppResult;
use crate::domain::models::{AppState, UserPreferences};
use crate::infrastructure::persistence::file_write::{read_json, write_json_atomic};
use crate::infrastructure::persistence::runtime_paths::RuntimePaths;

#[derive(Debug, Deserialize, Serialize, Default)]
#[serde(rename_all = "camelCase")]
struct AppStateFile {
    last_login_time: Option<DateTime<Local>>,
    last_quota_refresh_time: Option<DateTime<Local>>,
    last_login_result: String,
    last_login_message: String,
    recent_account_ids: Vec<String>,
    minimize_to_tray_on_close: bool,
    launch_on_startup: bool,
    auto_switch_account_on_traffic_exhausted: bool,
    migration_version: u32,
}

#[derive(Clone)]
pub struct AppStateRepository {
    paths: RuntimePaths,
}

impl AppStateRepository {
    pub fn new(paths: RuntimePaths) -> Self {
        Self { paths }
    }

    pub fn load_state(&self) -> AppResult<AppState> {
        let file = self.load_file()?;
        Ok(AppState {
            last_login_time: file.last_login_time,
            last_quota_refresh_time: file.last_quota_refresh_time,
            last_login_result: if file.last_login_result.trim().is_empty() {
                "未执行".to_string()
            } else {
                file.last_login_result
            },
            last_login_message: if file.last_login_message.trim().is_empty() {
                "-".to_string()
            } else {
                file.last_login_message
            },
            recent_account_ids: normalize_recent_account_ids(file.recent_account_ids),
            migration_version: file.migration_version,
        })
    }

    pub fn load_preferences(&self) -> AppResult<UserPreferences> {
        let file = self.load_file()?;
        Ok(UserPreferences {
            minimize_to_tray_on_close: file.minimize_to_tray_on_close,
            launch_on_startup: file.launch_on_startup,
            auto_switch_account_on_traffic_exhausted: file.auto_switch_account_on_traffic_exhausted,
        })
    }

    pub fn save_state(&self, state: &AppState) -> AppResult<()> {
        let mut file = self.load_file()?;
        file.last_login_time = state.last_login_time;
        file.last_quota_refresh_time = state.last_quota_refresh_time;
        file.last_login_result = state.last_login_result.clone();
        file.last_login_message = state.last_login_message.clone();
        file.recent_account_ids = normalize_recent_account_ids(state.recent_account_ids.clone());
        file.migration_version = state.migration_version;
        self.save_file(&file)
    }

    pub fn save_preferences(&self, preferences: &UserPreferences) -> AppResult<()> {
        let mut file = self.load_file()?;
        file.minimize_to_tray_on_close = preferences.minimize_to_tray_on_close;
        file.launch_on_startup = preferences.launch_on_startup;
        file.auto_switch_account_on_traffic_exhausted =
            preferences.auto_switch_account_on_traffic_exhausted;
        self.save_file(&file)
    }

    pub fn mark_account_used(&self, account_id: &str) -> AppResult<Vec<String>> {
        let mut state = self.load_state()?;
        let account_id = account_id.trim();
        if account_id.is_empty() {
            return Ok(state.recent_account_ids);
        }
        state.recent_account_ids.retain(|item| item != account_id);
        state.recent_account_ids.insert(0, account_id.to_string());
        state.recent_account_ids = normalize_recent_account_ids(state.recent_account_ids);
        self.save_state(&state)?;
        Ok(state.recent_account_ids)
    }

    pub fn prune_recent_account_ids(
        &self,
        valid_account_ids: &std::collections::HashSet<String>,
    ) -> AppResult<Vec<String>> {
        let mut state = self.load_state()?;
        state
            .recent_account_ids
            .retain(|account_id| valid_account_ids.contains(account_id));
        self.save_state(&state)?;
        Ok(state.recent_account_ids)
    }

    fn load_file(&self) -> AppResult<AppStateFile> {
        read_json(&self.paths.app_state_path())
    }

    fn save_file(&self, file: &AppStateFile) -> AppResult<()> {
        write_json_atomic(&self.paths.app_state_path(), file)
    }
}

fn normalize_recent_account_ids(raw: Vec<String>) -> Vec<String> {
    let mut seen = std::collections::HashSet::new();
    let mut result = Vec::new();
    for item in raw {
        let id = item.trim().to_string();
        if !id.is_empty() && seen.insert(id.clone()) {
            result.push(id);
        }
        if result.len() >= 20 {
            break;
        }
    }
    result
}
